_posts/2016-10-12-edsls-part-2.md

```diff
@@ -0,0 +1,413 @@
+    State.modify(_ - i)
+  def exists(i: Int): State[Set[Int], Boolean] =
+    State.inspect(_(i))
+}
+```
+
+```scala
+val state = program[State[Set[Int], ?]](ScalaSet)
+// state: cats.data.StateT[cats.Eval,scala.collection.immutable.Set[Int],Boolean] = cats.data.StateT@5718268a
+
+state.run(Set.empty).value
+// res4: (scala.collection.immutable.Set[Int], Boolean) = (Set(10),true)
+```
+
+Note that using `program` does not require any context-specific knowledge -
+we could just as well define another interpreter, perhaps one that does
+logging, and interpret the program that way.
+
+```scala
+import cats.data.StateT
+import scala.concurrent.{ExecutionContext, Future}
+
+// Asynchronous state
+def AsyncSet(implicit ec: ExecutionContext): SetLang[StateT[Future, Set[Int], ?]] =
+  new SetLang[StateT[Future, Set[Int], ?]] {
+    val monad = Monad[StateT[Future, Set[Int], ?]]
+    def add(i: Int): StateT[Future, Set[Int], Unit] =
+      StateT.modify(_ + i)
+    def remove(i: Int): StateT[Future, Set[Int], Unit] =
+      StateT.modify(_ - i)
+    def exists(i: Int): StateT[Future, Set[Int], Boolean] =
+      StateT.inspect(_(i))
+  }
+```
+
+```scala
+// No changes to `program` required
+val result = program(AsyncSet(ExecutionContext.global))
+// result: cats.data.StateT[scala.concurrent.Future,scala.collection.immutable.Set[Int],Boolean] = cats.data.StateT@24ca528d
+```
+
+Our `SetLang` captures the *structure* of a computation, but leaves open
+its *interpretation*.
+
+# Monad transformers and classes
+
+Our set language started with simple methods and is now fully generic and
+composable. As it turns out our `SetLang` encoding often referred to as
+[MTL-style][mtl].
+
+## Monads in monads
+
+Among the motivations for monad classes is to remove the need to specify
+monad transformer stacks. The following example is adapted from
+[Functional Programming with Overloading and Higher-Order Polymorphism][mtlPaper]
+by Professor Mark P. Jones.
+
+Consider a program that is open to failure and computes with some state. This
+suggests a combinator of `Either` and `State`, which in turn suggests their
+corresponding monad transformers. All that is left is to decide which
+transformer to use.
+
+```scala
+type App1[A] = EitherT[State[S, ?], Error, A]
+            // State[S, Either[Error, A]]
+            // S => (S, Either[Error, A])
+
+type App2[A] = StateT[Either[Error, ?], S, A]
+            // S => Either[Error, (S, A)]
+```
+
+While `App1` and `App2` are both valid compositions, the
+semantics of the compositions differ. `App1` describes a program where
+the computation of a *value* at each transition may fail whereas `App2`
+describes a program where the *entire* transition may fail.
+
+We can abstract away the difference by creating a type class which provides
+the relevant operations we need.
+
+```scala
+trait MonadError[F[_], E] {
+  def monad: Monad[F]
+
+  def raiseError[A](e: E): F[A]
+  def handleErrorWith[A](fa: F[A])(f: E => F[A]): F[A]
```

Comment:
A simpler encoding of `handleErrorWith` may be

```scala
[A](fa: F[A]): F[Either[E, A]]
```

Assuming `Functor[F]`, it's equally powerful.